### ISSUE TITLE
Haproxy 2.6 - FATAL: bug condition "fdtab[fd].owner != ((void *)0)"

### ISSUE THREAD

**Issue by hont512** [haproxy/haproxy#1777]
### Detailed Description of the Problem

With latest commit #db43010 (from native 2.6 repo) when I start Haproxy using systemd-service I get:
```
02:42:21 systemd  Starting HAProxy Load Balancer...
02:42:24 haproxy  [NOTICE]   (1974303) : New worker (1974305) forked
02:42:24 haproxy  [NOTICE]   (1974303) : Loading success.
02:42:24 haproxy  FATAL: bug condition "fdtab[fd].owner != ((void *)0)" matched at include/haproxy/fd.h:334
02:42:24 haproxy    call trace(7):
02:42:24 haproxy    | 0x55d5a1d0723e [c6 04 25 01 00 00 00 00]: main-0x8b2
02:42:24 haproxy    | 0x55d5a1e448c5 [85 c0 0f 84 59 02 00 00]: main+0x13cdd5
02:42:24 haproxy    | 0x55d5a1d096dc [31 c0 e8 1d 0f 1a 00 31]: main+0x1bec/0x2dda
02:42:24 kernel [530254.372588] haproxy[1974305]: segfault at 1 ip 000055d5a1d0723e sp 00007ffe22829100 error 6 in haproxy[55d5a1d02000+1c7000]
02:42:24 kernel [530254.372913] Code: e8 77 ff 15 00 e8 02 f7 15 00 c6 04 25 01 00 00 00 00 ba 03 00 00 00 48 8d 35 7e 36 1c 00 31 ff e8 57 ff 15 00 e8 e2 f6 15 00 <c6> 04 25 01 00 00 00 00 ba 03 00 00 00 48 8d 35 76 9d 1c 00 31 ff
02:42:24 haproxy  [NOTICE]   (1974303) : haproxy version is 2.6.1-db4301-23
02:42:24 systemd  Started HAProxy Load Balancer.
02:42:24 haproxy  [NOTICE]   (1974303) : path to executable is /usr/sbin/haproxy
02:42:24 haproxy  [ALERT]    (1974303) : Current worker (1974305) exited with code 139 (Segmentation fault)
02:42:24 haproxy  [ALERT]    (1974303) : exit-on-failure: killing every processes with SIGTERM
02:42:24 haproxy  [WARNING]  (1974303) : All workers exited. Exiting... (139)
02:42:24 systemd  haproxy.service: Main process exited, code=exited, status=139/n/a
02:42:24 systemd  haproxy.service: Failed with result 'exit-code'.
``` 

If I use 2.6 -> #7a792bd (just before commit "MINOR: fd: Add BUG_ON checks on fd_insert()") - Haproxy starts and works OK

### Expected Behavior

No error

### Steps to Reproduce the Behavior

Start

### What is your configuration?

```haproxy
I have large config with 500+ backends, multiple frontends and lua-items. And running on customized kernel.

Also, I'm just started to test move from 2.0 to 2.6, so some old ways might be used

However, I've tried already to disable lua and other suspicious things, no luck


global
lua-load /etc/haproxy/haproxy.lua
daemon
master-worker no-exit-on-failure
user haproxy
group haproxy
maxconn 27000
spread-checks 3
tune.bufsize 32768
tune.ssl.default-dh-param 2048
tune.ssl.maxrecord 4257
tune.h2.max-concurrent-streams 256
ssl-default-bind-options no-sslv3
ssl-default-server-options no-sslv3
defaults
default_backend backs
errorfile 408 /dev/null
errorfile 503 /etc/haproxy/errors/503.http
errorfile 403 /etc/haproxy/errors/403.http
option allbackups
option tcp-smart-accept
option tcp-smart-connect
http-reuse safe
mode http
log global
option abortonclose
option dontlognull
timeout connect 4s
timeout client 300s
timeout server 300s
timeout server-fin 30s
timeout client-fin 30s
timeout tunnel 50s
timeout http-request 30s
timeout queue 30s
timeout tarpit 1m
backlog 10000
timeout check 40
http-check expect status 718
default-server verify none inter 500 fastinter 1 fall 3 rise 2 maxqueue 1
option redispatch
retries 3
```
```


### Output of `haproxy -vv`

```plain
HAProxy version 2.6.1-db4301-23 2022/07/08 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.1.html
Running on: Linux 5.15.52 #1 SMP Mon Jul 4 19:37:49 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = native
  CC      = cc
  CFLAGS  = -O2 -march=native -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_THREAD=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_TFO=1 USE_NS=1 USE_SYSTEMD=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
Running on OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

**Comment by wtarreau** [comment]
Interesting. Could you please resolve the address using:
```
$ addr2line -e haproxy <<< 0x55d5a1d0723e
```

Hmmm I guess I got it. Apparently we've backported 36d9097cf ("MINOR: fd: Add BUG_ON checks on fd_insert()") without 34aae2fd1 ("MEDIUM: mworker: set the iocb of the socketpair without using fd_insert()"), and while that last one isn't originally a fix, it fixes a practice that became a misuse of the internal FD API and that now happens only with accidental reuse of an FD.

Given that you were able to bisect the commit ID, I guess you'd be at ease with backporting 34aae2fd1 ? It should solve it for you, I'll backport it to 2.6 and add a note about that one.

Thank you!

**Comment by hont512** [comment]
```
/usr/sbin # addr2line -e haproxy <<< 0x55d5a1d0723e
??:0
```

**Comment by hont512** [comment]
> Given that you were able to bisect the commit ID, I guess you'd be at ease with backporting [34aae2f](https://github.com/haproxy/haproxy/commit/34aae2fd1293872345056a32dec8047a67c3d852) ? It should solve it for you, I'll backport it to 2.6 and add a note about that one.

Confirm
#db43010 + that fd_insert change = no more infinity process dying

**Comment by wtarreau** [comment]
Thank you!

**Comment by wtarreau** [comment]
I've marked it as fix pending for 2.6 so that we don't forget!

**Comment by wlallemand** [comment]
I backported it to 2.6.
